specifyweb/specify/calculated_fields.py

```diff
@@ -19,94 +19,117 @@ def calculate_extra_fields(obj, data: Dict[str, Any]) -> Dict[str, Any]:
+
+    elif isinstance(obj, get_model("Accession")):
+        Preparation = get_model("Preparation")
+        preparations = Preparation.objects.filter(collectionobject__accession=obj)
+        preparationCount = preparations.count()
+        totalCountAmt = preparations.aggregate(total=Sum("countamt"))["total"] or 0
+
+        actualTotalCountAmt = 0
+        for prep in preparations:
+            giftpreparation_quantity = prep.giftpreparations.aggregate(total=Sum("quantity"))["total"] or 0
+            exchangeoutprep_quantity = prep.exchangeoutpreps.aggregate(total=Sum("quantity"))["total"] or 0
+            disposalpreparation_quantity = prep.disposalpreparations.aggregate(total=Sum("quantity"))["total"] or 0
+
+            countamt = obj.countamt or 0
+            available = max(0, countamt - giftpreparation_quantity - exchangeoutprep_quantity - disposalpreparation_quantity)
+            actualTotalCountAmt += available
+
         extra["actualTotalCountAmt"] = int(actualTotalCountAmt)
         extra["totalCountAmt"] = int(totalCountAmt)
         extra["preparationCount"] = preparationCount
-        extra.update(obj.collectionobjects.aggregate(collectionObjectCount=Count('id')))
+        extra.update(obj.collectionobjects.aggregate(collectionObjectCount=Count("id")))
+
+    elif isinstance(obj, get_model("Disposal")):
+        extra["totalPreps"] = obj.disposalpreparations.count()
+        extra["totalItems"] = obj.disposalpreparations.aggregate(total=Sum("quantity"))["total"] or 0
+
+    elif isinstance(obj, get_model("Gift")):
+        extra["totalPreps"] = obj.giftpreparations.count()
+        extra["totalItems"] = obj.giftpreparations.aggregate(total=Sum("quantity"))["total"] or 0
+
+    elif isinstance(obj, get_model("ExchangeOut")):
+        extra["totalPreps"] = obj.exchangeoutpreps.count()
+        extra["totalItems"] = obj.exchangeoutpreps.aggregate(total=Sum("quantity"))["total"] or 0
+
+    elif isinstance(obj, get_model("Deaccession")):
+        Disposal = get_model("Disposal")
+        Exchangeout = get_model("ExchangeOut")
+        Gift = get_model("Gift")
+        disposals = Disposal.objects.filter(deaccession=obj)
+        gifts = Gift.objects.filter(deaccession=obj)
+        exchangeouts = Exchangeout.objects.filter(deaccession=obj)
+        # sum up all totalPreps of disposals, exchangeouts and gifts
+        extra["totalPreps"] = sum(obj["totalPreps"] for obj in disposals) + \
+            sum(obj["totalPreps"] for obj in exchangeouts) + \
+                sum(obj["totalPreps"] for obj in gifts)
+
+        # sum up all totalItems of disposals, exchangeouts and gifts
+        extra["totalItems"] = sum(obj["totalItems"] for obj in disposals if obj["totalItems"] is not None) + \
+            sum(obj["totalItems"] for obj in exchangeouts if obj["totalItems"] is not None) + \
+                sum(obj["totalItems"] for obj in gifts if obj["totalItems"] is not None)
```

Comment:
thanks for that change. does django allow aggregating here? for loop can affect performance, you are fetching all objects, all of their fields, but just looking at two of them. Maybe we should add a rule "every time you fetch an entire object without a good reason, you donate a dollar to a jar for getting faster hardware"

Reply:
lol, fixed. I abstracted a lot of the code in this file to reduce code repetition